mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocSaver.java

```diff
@@ -288,59 +293,34 @@ private void updateDocs(CRUDOperationContext ctx,
                         try {
                             ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.PRE_CRUD_UPDATE_DOC, ctx, doc.inputDoc);
                             merge.merge(doc.oldDoc, doc.newDoc);
-                            Translator.populateDocHiddenFields(doc.newDoc, md);
+                            Translator.populateCaseInsensitiveHiddenFields(doc.newDoc, md);
                             updateAttemptList.add(doc);
                         } catch (Exception e) {
                             doc.inputDoc.addError(Error.get("update", MongoCrudConstants.ERR_TRANSLATION_ERROR, e));
                         }
                     } else {
                         doc.inputDoc.addError(Error.get("update",
-                                CrudConstants.ERR_NO_FIELD_UPDATE_ACCESS, paths.toString()));
+                                                        CrudConstants.ERR_NO_FIELD_UPDATE_ACCESS, paths.toString()));
                     }
                 }
                 LOGGER.debug("After checks and merge, updating {} docs", updateAttemptList.size());
                 if (!updateAttemptList.isEmpty()) {
-                    BulkWriteOperation bw = collection.initializeUnorderedBulkOperation();
+                    BulkUpdateCtx updateCtx=new BulkUpdateCtx(collection);
                     for (DocInfo doc : updateAttemptList) {
-                        bw.find(new BasicDBObject("_id", doc.oldDoc.get("_id"))).replaceOne(doc.newDoc);
-                        doc.inputDoc.setCRUDOperationPerformed(CRUDOperation.UPDATE);
+                        BulkUpdateCtx.UpdateDoc updateDoc=new BulkUpdateCtx.UpdateDoc(doc.oldDoc,doc.inputDoc);
+                        updateCtx.addDoc(updateDoc,doc.newDoc);
                     }
-                    try {
-                        if (writeConcern == null) {
-                            LOGGER.debug("Bulk updating docs");
-                            bw.execute();
-                        }
-                        else {
-                            LOGGER.debug("Bulk updating docs with writeConcern={} from execution", writeConcern);
-                            bw.execute(writeConcern);
-                        }
-                    } catch (BulkWriteException bwe) {
-                        LOGGER.error("Bulk write exception", bwe);
-                        handleBulkWriteError(bwe.getWriteErrors(), "update", updateAttemptList);
-                    } catch (RuntimeException e) {
-                    } finally {
-                        for (DocInfo doc : updateAttemptList) {
-                            if (!doc.inputDoc.hasErrors()) {
-                                ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_UPDATE_DOC, ctx, doc.inputDoc);
-                            }
+                    updateCtx.execute(writeConcern);
```

Comment:
For save requests, it's interesting that we're detecting concurrent mods too. Is it because it's possible identifying values could have changed between when we retrieved the document to save over and when we actually do the save? Are identity fields allowed to be mutable?

I'm trying to decide if clients issuing a save request would want this behavior or not (and if they do, maybe they really should be doing an update). I guess I expect a save to potentially blow away other concurrent updates since I'm passing in the whole entity.